compiler/src/iree/compiler/Codegen/SPIRV/KernelConfig.cpp

```diff
@@ -48,6 +49,37 @@ bool isMatmulOrBatchMatmul(linalg::LinalgOp linalgOp) {
          llvm::is_contained({2u, 3u}, linalgOp.getNumParallelLoops());
 }
 
+static bool containsUnsupportedOps(Block &block) {
+  bool hasUnsupportedOps = false;
+  block.walk([&](Operation *op) {
+    if (isa<arith::ExtFOp>(op)) {
+      hasUnsupportedOps = true;
+      return WalkResult::interrupt();
+    }
+    return WalkResult::advance();
+  });
+  return hasUnsupportedOps;
+}
+
+// Check if the given linalg op is fused with another op that may block
+// multi-buffering
```

Comment:
I found that if the matmul is fused with an elementwise extf it will result in an alloc that multi-buffering can't find a loop that dominate the uses of it. Something like this
```
builtin.module {
  func.func @forward(%arg0: tensor<1x4096x512xf16>, %arg1: tensor<1x512x4096xf16>) -> tensor<1x4096x4096xf32> {
    %cst = arith.constant 0.000000e+00 : f16
    %5 = tensor.empty() : tensor<1x4096x4096xf32>
    %6 = tensor.empty() : tensor<1x4096x4096xf16>
    %7 = linalg.fill ins(%cst : f16) outs(%6 : tensor<1x4096x4096xf16>) -> tensor<1x4096x4096xf16>
    %8 = linalg.batch_matmul ins(%arg0, %arg1 : tensor<1x4096x512xf16>, tensor<1x512x4096xf16>) outs(%7 : tensor<1x4096x4096xf16>) -> tensor<1x4096x4096xf16>
    %9 = linalg.generic {indexing_maps = [affine_map<(d0, d1, d2) -> (d0, d1, d2)>, affine_map<(d0, d1, d2) -> (d0, d1, d2)>], iterator_types = ["parallel", "parallel", "parallel"]} ins(%8 : tensor<1x4096x4096xf16>) outs(%5 : tensor<1x4096x4096xf32>) {
    ^bb0(%in: f16, %out: f32):
      %10 = arith.extf %in : f16 to f32
      linalg.yield %10 : f32
    } -> tensor<1x4096x4096xf32>
    return %9 : tensor<1x4096x4096xf32>
  }
}

```
I can turn this into a test if that helps.

Reply:
I guess it's because `arith.extf` extends element bitwidth so it requires a larger output buffer output (`tensor<1x4096x4096xf32>`) than the matmul generates (`tensor<1x4096x4096xf16>`). Bufferization will create an intermediate buffer for it then. This seems to be a case we should fix bufferization for. Though then this is not specific to `arith.extf`; I suspect `arith.extsi` or similar ones will trigger this too.

Yes having such a test case would be helpful. Together, can we change this to check `OpTrait::hasElementwiseMappableTraits` and 1 operand and 1 result, but operand and result type have different bitwidth type? Some comments explaining this would also be great.

Reply:
> I guess it's because arith.extf extends element bitwidth so it requires a larger output buffer.

Yeah that is what I found, but perhaps just checking bitwidth makes more sense then, although I suspect there could still be cases missing. Namely, there are examples where I see a fusion with truncf that doesn't fail multi-buffering but I didn't take enough time to figure out why, although it would make sense that extf/si would come up with an additional buffer.

Reply:
I had to add something similar back in because the total shared memory usage wasn't being calculated properly. If there is a suggestion for a better way to handle this that would be best, otherwise I left it as a TODO that can be fixed with bufferization or by properly computing the shared memory usage.